### PR TITLE
Run 5188/part 1

### PR DESCRIPTION
**Pull request by pbaize**
#### Description of Change
Extract webcontents methods from window.js
First PR to allow code sharing with browserview

#### Checklist

- [x] PR description included and stakeholders cc'd
- [x] `npm test` passes
- [x] PR title starts with the JIRA ticket [pull request process](https://github.com/openfin/Internal-Wiki/wiki/Pull-Request-Process)

#### Release Notes

Notes: na

### CODE REVIEW

File: src/browser/api/webcontents.ts [hunk 1]
```diff
@@ -0,0 +1,56 @@
+import { WebContents } from 'electron';
+import * as url from 'url';
+
+
+export function executeJavascript(webContents: WebContents, code: string, callback: (e: any, result: any) => void): void {
```
Comment (whyn07m3) [inline src/browser/api/webcontents.ts]: `e` here is always undefined here instead of `any`?

File: src/browser/api/webcontents.ts [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+    webContents.executeJavaScript(code, true, (result) => {
```
Comment (whyn07m3) [inline src/browser/api/webcontents.ts]: can/should we do `await` here, [electron docs](https://github.com/openfin/runtime/blob/d648b1832101766bc369c7cb6b42920b95886c8e/docs/api/web-contents.md#contentsexecutejavascriptcode-usergesture-callback)?

File: src/browser/api/webcontents.ts [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+        callback(undefined, result);
+    });
+}
+
+export function getAbsolutePath(webContents: WebContents, path: string) {
```
Comment (whyn07m3) [inline src/browser/api/webcontents.ts]: same for other functions: do we need to specify return types?

File: src/browser/api/window.js [hunk 4]
```diff
@@ -1769,15 +1755,12 @@ Window.authenticate = function(identity, username, password, callback) {
 
 Window.getZoomLevel = function(identity, callback) {
     let browserWindow = getElectronBrowserWindow(identity, 'get zoom level for');
-
-    browserWindow.webContents.getZoomLevel(callback);
+    WebContents.getZoomLevel(browserWindow.webContents, callback);
```
Comment (whyn07m3) [inline src/browser/api/window.js]: doesn't this need to `return`?

File: src/browser/api/webcontents.ts [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+    return {
+        canNavigateBack: webContents.canGoBack(),
+        canNavigateForward: webContents.canGoForward(),
+        title: webContents.getTitle(),
+        url: webContents.getURL()
+    };
+}
+
+export function getZoomLevel(webContents: WebContents, callback: (zoomLevel: number) => void) {
+    webContents.getZoomLevel(callback);
```
Comment (whyn07m3) [inline src/browser/api/webcontents.ts]: doesn't this need to `return`?

File: src/browser/api/webcontents.ts [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+export function getInfo(webContents: WebContents) {
+    return {
+        canNavigateBack: webContents.canGoBack(),
+        canNavigateForward: webContents.canGoForward(),
+        title: webContents.getTitle(),
+        url: webContents.getURL()
+    };
+}
+
+export function getZoomLevel(webContents: WebContents, callback: (zoomLevel: number) => void) {
```
Comment (whyn07m3) [inline src/browser/api/webcontents.ts]: can/should we make this function `async` and handle callback here? (thinking about the future...)

File: src/browser/api/webcontents.ts [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+export function reload(webContents: WebContents, ignoreCache: boolean = false) {
+    if (!ignoreCache) {
+        webContents.reload();
+    } else {
+        webContents.reloadIgnoringCache();
+    }
+}
+
+export function setZoomLevel(webContents: WebContents, level: number) {
+    // webContents.setZoomLevel(level); // zooms all windows loaded from same domain
```
Comment (whyn07m3) [inline src/browser/api/webcontents.ts]: will we need this corpse in the future?

**Comment by whyn07m3** [comment]
Please, resolve conflicts, so that we can review final version

File: src/browser/api/webcontents.ts (hunk 7 above)
Comment (pbaize) [reply src/browser/api/webcontents.ts]: Opened a ticket to investigate that

**Comment by pbaize** [comment]
Something got messed up in the rebase, hold on review

File: src/browser/api/webcontents.ts (hunk 6 above)
Comment (pbaize) [reply src/browser/api/webcontents.ts]: I'd rather do that in a follow up pr, keep this as close to the original window.js as we can for now

File: src/browser/api/webcontents.ts (hunk 5 above)
Comment (pbaize) [reply src/browser/api/webcontents.ts]: Not with the callback, this code will be dead in part-2 of this pr anyways

File: src/browser/api/webcontents.ts (hunk 2 above)
Comment (pbaize) [reply src/browser/api/webcontents.ts]: Rather do it in a follow up, keep this closer to current code

File: src/browser/api/webcontents.ts [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+
+export function stopNavigation(webContents: Electron.WebContents) {
+    webContents.stop();
+}
+
+function createNavigationEndPromise(webContents: Electron.WebContents): Promise<void> {
+    return new Promise((resolve, reject) => {
+        const chromeErrCodesLink = 'https://cs.chromium.org/chromium/src/net/base/net_error_list.h';
+        // tslint:disable-next-line:no-empty
+        let didSucceed = () => { };
```
Comment (whyn07m3) [inline src/browser/api/webcontents.ts]: isn't `didSucceed` always gonna be what's being assigned on line #83?

File: src/browser/api/webcontents.ts [hunk 9]
```diff
@@ -0,0 +1,90 @@
...
+        canNavigateBack: webContents.canGoBack(),
+        canNavigateForward: webContents.canGoForward(),
+        title: webContents.getTitle(),
+        url: webContents.getURL()
+    };
+}
+
+export function getAbsolutePath(webContents: Electron.WebContents, path: string) {
+    const windowURL = webContents.getURL();
+
```
Comment (whyn07m3) [inline src/browser/api/webcontents.ts]: :neckbeard: extra empty line?

File: src/browser/api/webcontents.ts [hunk 10]
```diff
@@ -0,0 +1,90 @@
...
+    return  url.resolve(windowURL, path);
+}
+
+export function navigate (webContents: Electron.WebContents, url: string) {
+    // todo: replace everything here with "return webContents.loadURL(url)" once we get to electron 5.*
```
Comment (whyn07m3) [inline src/browser/api/webcontents.ts]: Good time to add a Jira ticket to clean it up

Comment (pbaize) [reply src/browser/api/webcontents.ts]: Already done

File: src/browser/api/webcontents.ts (hunk 8 above)
Comment (pbaize) [reply src/browser/api/webcontents.ts]: Yeah, just being paranoid, will fix

File: src/browser/api/webcontents.ts [hunk 11]
```diff
@@ -0,0 +1,90 @@
...
+    }
+    const navigationEnd = createNavigationEndPromise(webContents);
+    webContents.goBack();
+    return navigationEnd;
+}
+
+export async function navigateForward (webContents: Electron.WebContents) {
+    if (!webContents.canGoForward()) {
+        const error = new Error('Cannot navigate forward');
+        throw error;
```
[the remark this answers is not in the document]
Comment (pbaize) [reply src/browser/api/webcontents.ts]: 👍

File: src/browser/api/webcontents.ts [hunk 12]
```diff
@@ -0,0 +1,90 @@
...
+    }
+    const navigationEnd = createNavigationEndPromise(webContents);
+    webContents.goForward();
+    return navigationEnd;
+}
+
+export function getZoomLevel(webContents: Electron.WebContents, callback: (zoomLevel: number) => void) {
+    webContents.getZoomLevel(callback);
```
Comment (tomer-openfin) [inline src/browser/api/webcontents.ts]: it would've been nice if this used a promise instead of a cb

Comment (pbaize) [reply src/browser/api/webcontents.ts]: That's not how the electron api works for this one

Comment (tomer-openfin) [reply src/browser/api/webcontents.ts]: I'm suggesting returning a promise that resolves when the call is done, like so:

> return new Promise(res => webContents.getZoomLevel(res));

Comment (pbaize) [reply src/browser/api/webcontents.ts]: I agree it's better, but I'm trying to stick as close to the window.js code as I can. It's already a pretty large change set and I want to keep it as safe as I can.

Comment (tomer-openfin) [reply src/browser/api/webcontents.ts]: makes sense.

**Comment by datamadic** [comment]
Looks good, discussed next steps and perhaps a part 3. Lets see how this one does as a build and merge part 2 right after tomorrow if it all goes well. Looking at the (now skewed) test results this looks fine as per best current canary run, hopefully we can run part 2 in a clean test env...

**Pull request merged**